pkg/api/universal/conversation.go

```diff
@@ -115,26 +110,89 @@ func (a *Universal) ConverseAlpha1(ctx context.Context, req *runtimev1pb.Convers
+func (a *Universal) ConverseStreamAlpha1(req *runtimev1pb.ConversationRequest, stream runtimev1pb.Dapr_ConverseStreamAlpha1Server) error {
+	// Validate component exists
+	if a.compStore.ConversationsLen() == 0 {
+		err := messages.ErrConversationNotFound
+		a.logger.Debug(err)
+		return err
+	}
+
+	component, ok := a.compStore.GetConversation(req.GetName())
+	if !ok {
+		err := messages.ErrConversationNotFound.WithFormat(req.GetName())
+		a.logger.Debug(err)
+		return err
+	}
+
+	if len(req.GetInputs()) == 0 {
+		err := messages.ErrConversationMissingInputs.WithFormat(req.GetName())
+		a.logger.Debug(err)
+		return err
+	}
 
-			response.Outputs = append(response.GetOutputs(), &runtimev1pb.ConversationResult{
-				Result:     res,
-				Parameters: o.Parameters,
-			})
+	// Create streaming pipeline
+	pipelineImpl := conversation.NewStreamingPipelineImpl(a.logger)
+
+	// prepare non-streaming scrubber in case PII scrubbing is enabled
+	var inputScrubber, outputScrubber piiscrubber.Scrubber
+
+	// Check if PII scrubbing is needed for inputs or outputs
+	needsInputPIIScrubbing := conversation.NeedsInputScrubber(req)
+	needsOutputPIIScrubbing := conversation.NeedsOutputScrubber(req)
+
+	if needsInputPIIScrubbing || needsOutputPIIScrubbing {
+		scrubber, err := piiscrubber.NewDefaultScrubber()
+		if err != nil {
+			return messages.ErrConversationScrubbingFailed.WithFormat(req.GetName(), err.Error())
+		}
+		if needsInputPIIScrubbing {
+			inputScrubber = scrubber
+		}
+		if needsOutputPIIScrubbing {
+			outputScrubber = scrubber
+			scrubberMiddleware, err := conversation.NewStreamingPIIScrubber()
+			if err != nil {
+				return fmt.Errorf("failed to create streaming PII scrubber middleware: %w", err)
+			}
+			pipelineImpl.AddMiddleware(scrubberMiddleware)
 		}
 	}
 
-	return response, nil
+	// Prepare component conversation componentRequest (similar to non-streaming version)
+	componentRequest, err := conversation.GetComponentRequest(req, inputScrubber)
+	if err != nil {
+		return err
+	}
+
+	// Process streaming request
+	ctx := stream.Context()
+	start := time.Now()
+	var usage *runtimev1pb.ConversationUsage
+	var retErr error // used in defer to track error
+	defer func() {
+		elapsed := diag.ElapsedSince(start)
+		var promptTokens, completionTokens int64
+		if usage != nil {
+			promptTokens = int64(usage.GetPromptTokens())         //nolint:gosec // Intentional use of int64 for backward compatibility
+			completionTokens = int64(usage.GetCompletionTokens()) //nolint:gosec // Intentional use of int64 for backward compatibility
+		}
+		diag.DefaultComponentMonitoring.ConversationInvoked(ctx, req.GetName(), retErr == nil, elapsed, diag.StreamingConversation, promptTokens, completionTokens)
+	}()
+
+	// we use non-streaming scrubber on the non-streamingoutput
+	// we get usage to use in monitoring
+	usage, retErr = pipelineImpl.ProcessStream(ctx, stream, component, componentRequest, req.GetName(), outputScrubber)
+	return retErr
```

Comment:
Please move the sending process into a new func. Assigning vars (`usage`) which are re-referenced in a `defer` mid function is confusing to read.